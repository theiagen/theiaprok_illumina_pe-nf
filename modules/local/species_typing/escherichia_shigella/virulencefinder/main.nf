process VIRULENCEFINDER {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/staphb/virulencefinder:2.0.4"

    input:
    tuple val(meta), path(assembly)
    val database
    val min_percent_coverage
    val min_percent_identity

    output:
    tuple val(meta), path("*_results_tab.tsv"), emit: virulence_report
    tuple val(meta), path("VIRULENCE_FACTORS.txt"), emit: virulence_factors
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def database_arg = database ? database: "virulence_ecoli"
    def coverage_arg = min_percent_coverage ? "-l ${min_percent_coverage}" : ""
    def identity_arg = min_percent_identity ? "-t ${min_percent_identity}" : ""
    
    """
    date | tee DATE

    mkdir tmp
    
    virulencefinder.py \\
        -i ${assembly} \\
        -o . \\
        -tmp tmp \\
        -x \\
        -d ${database_arg} \\
        ${coverage_arg} \\
        ${identity_arg} \\
        ${args}
    
    # rename output files
    mv results_tab.tsv ${prefix}_results_tab.tsv

    # parse 2nd column (Virulence factor) of _tab.tsv file into a comma-separated string
    tail -n+2 ${prefix}_results_tab.tsv | awk '{print \$2}' | uniq | paste -s -d, - | tee VIRULENCE_FACTORS.txt

    # if virulencefinder fails/no hits are found, the results_tab.tsv file will still exist but only be the header
    # check to see if VIRULENCE_FACTORS is just whitespace
    # if so, say that no virulence factors were found instead
    if ! grep -q '[^[:space:]]' VIRULENCE_FACTORS.txt ; then 
      echo "No virulence factors found" | tee VIRULENCE_FACTORS.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        virulencefinder: \$(virulencefinder.py --version 2>&1 | grep -oP 'VirulenceFinder \\K[0-9.]+' || echo "2.0.4")
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_results_tab.tsv
    echo "No virulence genes detected" > VIRULENCE_FACTORS.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        virulencefinder: 2.0.4
    END_VERSIONS
    """
}