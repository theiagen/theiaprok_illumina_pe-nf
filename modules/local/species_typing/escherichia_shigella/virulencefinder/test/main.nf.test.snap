{
    "virulencefinder - test": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_results_tab.tsv:md5,bcb7f27deceee8ba2a356d38c01624e8"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "VIRULENCE_FACTORS.txt:md5,3367e51b942ab7c2ea68321397919c7d"
                ]
            ],
            [
                "versions.yml:md5,5b285e7bcfd868c47d3f3516911d854d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T19:53:34.991305795"
    },
    "virulencefinder - test - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_results_tab.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "VIRULENCE_FACTORS.txt:md5,a162cc1a4b531a49f95c357bfd360920"
                    ]
                ],
                "2": [
                    "versions.yml:md5,5b285e7bcfd868c47d3f3516911d854d"
                ],
                "versions": [
                    "versions.yml:md5,5b285e7bcfd868c47d3f3516911d854d"
                ],
                "virulence_factors": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "VIRULENCE_FACTORS.txt:md5,a162cc1a4b531a49f95c357bfd360920"
                    ]
                ],
                "virulence_report": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_results_tab.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-24T19:53:47.309303378"
    }
}